core/profiling/eventpipe/src/CorProfiler.cpp

```diff
@@ -0,0 +1,781 @@
+{
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CorProfiler::DynamicMethodUnloaded(FunctionID functionId)
+{
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CorProfiler::EventPipeEventDelivered(
+    EVENTPIPE_PROVIDER provider,
+    DWORD eventId,
+    DWORD eventVersion,
+    ULONG cbMetadataBlob,
+    LPCBYTE metadataBlob,
+    ULONG cbEventData,
+    LPCBYTE eventData,
+    LPCGUID pActivityId,
+    LPCGUID pRelatedActivityId,
+    ThreadID eventThread,
+    ULONG numStackFrames,
+    UINT_PTR stackFrames[])
+{
+    String name = GetOrAddProviderName(provider);
+    
+    EventPipeMetadataInstance metadata = GetOrAddMetadata(metadataBlob, cbMetadataBlob);
+    EventPipeEventPrinter printer;
+    printer.PrintEvent(name.ToNativeStr(),
+                       metadata,
+                       eventData,
+                       cbEventData,
+                       pActivityId,
+                       pRelatedActivityId,
+                       eventThread,
+                       stackFrames,
+                       numStackFrames);
+    return S_OK;
+}
+
+HRESULT CorProfiler::EventPipeProviderCreated(EVENTPIPE_PROVIDER provider)
+{
+    String name = GetOrAddProviderName(provider);
+    wprintf(L"CorProfiler::EventPipeProviderCreated provider=%s\n", name.ToCStr());
+
+    // Add all events from any new provider to the session
+    COR_PRF_EVENTPIPE_PROVIDER_CONFIG providerConfig = { name.ToNativeStr(), 0xFFFFFFFFFFFFFFFF, 5, NULL };
+    HRESULT hr = _pCorProfilerInfo12->EventPipeAddProviderToSession(_session, providerConfig);
+    if (FAILED(hr))
+    {
+        printf("EventPipeAddProviderToSession failed with hr=0x%x\n", hr);
+        return hr;
+    }
+
+    return S_OK;
+}
+
+String CorProfiler::GetOrAddProviderName(EVENTPIPE_PROVIDER provider)
+{
+    lock_guard<mutex> guard(_cacheLock);
+
+    auto it = _providerNameCache.find(provider);
+    if (it == _providerNameCache.end())
+    {
+        WCHAR nameBuffer[LONG_LENGTH];
+        ULONG nameCount;
+        HRESULT hr = _pCorProfilerInfo12->EventPipeGetProviderInfo(provider,
+                                                                   LONG_LENGTH,
+                                                                   &nameCount,
+                                                                   nameBuffer);
+        if (FAILED(hr))
+        {
+            printf("EventPipeGetProviderInfo failed with hr=0x%x\n", hr);
+            return WCHAR("GetProviderInfo failed");
+        }
+
+        _providerNameCache.insert({provider, String(nameBuffer)});
+
+        it = _providerNameCache.find(provider);
+        assert(it != _providerNameCache.end());
+    }
+
+    return it->second;
+}
+
+EventPipeMetadataInstance CorProfiler::GetOrAddMetadata(LPCBYTE pMetadata, ULONG cbMetadata)
+{
+    // TODO: holding the lock while parsing metdata is not the best plan. Metadata parsing
```

Comment:
I was just being lazy and planning on leaving the locking up to whoever was using the sample. Since you wrote the thread safe map code for me I'll just use that and remove the todo